p2panda-js/src/session.ts

```diff
@@ -0,0 +1,278 @@
+   * @returns Session
+   */
+  schema(val: string): Session {
+    this._schema = val;
+    return this;
+  }
+
+  __keyPair: KeyPair | null = null;
+
+  get _keyPair(): KeyPair {
+    if (!this.__keyPair) {
+      throw new Error(
+        'Requires a signing key pair. Configure a key pair with ' +
+          '`session.keyPair()` or with the `options` parameter on methods.',
+      );
+    }
+    return this._keyPair;
+  }
+
+  set _keyPair(val: KeyPair) {
+    this.__keyPair = val;
+  }
+
+  /**
+   * Set a fixed key pair for this session, which will be used by methods unless
+   * a different key pair is configured through their `options` parameters.
+   *
+   * @param val key pair instance generated using the `KeyPair` class.
+   * @returns key pair instance
+   */
+  keyPair(val: KeyPair): Session {
+    this._keyPair = val;
+    return this;
+  }
+
+  // Load and return the WebAssembly p2panda library.
+  //
+  // Always await this function before using `this.p2panda`. Unfortunately this
+  // cannot be handled in the constructor as the contructor cannot be async.
+  async loadWasm(): Promise<P2Panda> {
+    if (this.p2panda == null) {
+      this.p2panda = await p2panda;
+
+      // I am removing this again because it breaks the node build but it may
+      // have to go back in.
+
+      // if (this.p2panda.default != null) {
+      //   log('fallback loader from `p2panda.default`');
+      //   // @ts-expect-error only applies for node context
+      //   this.p2panda = await p2panda.default;
+      // } else {
+      //   log('loaded wasm lib');
+      // }
+      log('initialized wasm lib');
+    } else {
+      log('access cached wasm lib');
+    }
+    return this.p2panda;
+  }
+
+  /**
+   * Return arguments for constructing the next entry given author and schema.
+   *
+   * This uses the cache set through `Session._setNextEntryArgs`.
+   *
+   * @param author public key of the author
+   * @param schema schema id
+   * @returns an `EntryArgs` object
+   */
+  async _getNextEntryArgs(author: string, schema: string): Promise<EntryArgs> {
+    if (!author || !schema)
+      throw new Error('Author and schema must be provided');
+    const cacheKey = `${author}/${schema}`;
+    const cachedValue = this.nextEntryArgs[cacheKey];
+    if (cachedValue) {
+      // use cache
+      delete this.nextEntryArgs[cacheKey];
+      log('call panda_getEntryArguments [cached]', cachedValue);
+      return cachedValue;
+    } else {
+      // do rpc call
+      const nextEntryArgs = await this.client.request({
+        method: 'panda_getEntryArguments',
+        params: { author, schema },
+      });
+      log('call panda_getEntryArguments', nextEntryArgs);
+      return nextEntryArgs;
+    }
+  }
+
+  /**
+   * Cache next entry args for a given author and schema
+   *
+   * @param author public key of the author
+   * @param schema schema id
+   * @param entryArgs an object with entry arguments
+   */
+  _setNextEntryArgs(
```

Comment:
I don't know TypeScript too well, but is there a way to protect access to these private methods by declaring them "private" or "protected" via TypeScript?

Reply:
Yes! I never used them.

Reply:
I have made some fields `private`, but this also prevents us from using them in other `p2panda-js` modules (e.g. `instance.ts`) so I have reverted the naming with leading underscore.

Reply:
Does `protected` do the job? Sounds a little like what you are describing